airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -441,6 +458,53 @@ def _checkpoint_mode(self) -> CheckpointMode:
         else:
             return CheckpointMode.FULL_REFRESH
 
+    @staticmethod
+    def _detect_stream_category(mappings_or_slices: Iterable[Union[Mapping[str, Any], StreamSlice]]) -> tuple[bool, bool]:
+        """
+        This is a bit of a crazy solution, but also the only way we can detect certain attributes about the stream since Python
+        streams do not follow consistent implementation patterns. We care about the following two attributes:
+        - is_substream: Helps to incrementally release changes since substreams w/ parents are much more complicated. Also
+          helps de-risk the release of changes that might impact all connectors
+        - uses_legacy_slice_format: Since the checkpoint reader must manage a complex state object, we opted to have it always
+          use the structured StreamSlice object. However, this requires backwards compatibility with Python sources that only
+          support the legacy mapping object
+
+        Both attributes can eventually be deprecated once stream's define this method deleted once substreams have been implemented and
+        legacy connectors all adhere to the StreamSlice object.
+        """
+
+        try:
+            next_slice = next(mappings_or_slices)
+            if isinstance(next_slice, StreamSlice) and next_slice == StreamSlice(partition={}, cursor_slice={}):
+                is_legacy_format = False
+                slice_has_value = False
+            elif next_slice == {}:
+                is_legacy_format = True
+                slice_has_value = False
+            elif isinstance(next_slice, StreamSlice):
+                is_legacy_format = False
+                slice_has_value = True
+            else:
+                is_legacy_format = True
+                slice_has_value = True
+        except StopIteration:
+            # If the stream has no slices, the format ultimately does not matter since no data will get synced. This is technically
+            # a valid case because it is up to the stream to define its slicing behavior
+            return False, False
+
+        if slice_has_value:
+            # If the first slice contained a partition value from the result of stream_slices(), this is a substream that might
+            # have multiple parent records to iterate over
+            return is_legacy_format, slice_has_value
+
+        try:
+            # If stream_slices() returns multiple slices, this is also a substream that can potentially generate empty slices
```

Comment:
how do you know this is a substream and not a legacy "date time based cursor"? [example from stripe](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connectors/source-stripe/source_stripe/streams.py#L272)

Reply:
Hmmm, this is a good point, but in a way it might sort of work already... but not in a way that is immediately intuitive. Right now we might be categorizing regular incremental streams where slices are time windows as substreams.

The main use of this logic is to determine whether or not we should remove/disable the RFR cursor on the HttpStream for substreams.

I'll investigate this a bit, but what might be a better categorization is `has_multiple_slices`. If a stream has multiple slices and is incremental, for an `HttpStream` which can only have an RFR cursor or no cursor, we still want to disable the cursor and defer to the `IncrementalCheckpointReader`.

And if a stream has multiple slices and is not incremental, then we are already gating this off so we should still be disabling the cursor which will then still defer to the `FullRefreshCheckpointReader` which will iterate over every parent slice.

I'll sketch this out, but I think this should work